tests/test_discovery.py

```diff
@@ -533,14 +538,148 @@ def test_api_endpoint_override_from_client_options_dict(self):
         discovery = open(datafile("plus.json")).read()
         api_endpoint = "https://foo.googleapis.com/"
         plus = build_from_document(
-            discovery, 
+            discovery,
             client_options={"api_endpoint": api_endpoint},
-            credentials=self.MOCK_CREDENTIALS
+            credentials=self.MOCK_CREDENTIALS,
         )
 
         self.assertEqual(plus._baseUrl, api_endpoint)
 
 
+class DiscoveryFromDocumentMutualTLS(unittest.TestCase):
+    MOCK_CREDENTIALS = mock.Mock(spec=google.auth.credentials.Credentials)
+    ADC_CERT_PATH = "adc_cert_path"
+    ADC_KEY_PATH = "adc_key_path"
+    ADC_PASSPHRASE = "adc_passphrase"
+    REGULAR_ENDPOINT = "https://www.googleapis.com/plus/v1/"
+    MTLS_ENDPOINT = "https://www.mtls.googleapis.com/plus/v1/"
+
+    def check_http_client_cert(self, resource, has_client_cert=False):
+        if isinstance(resource._http, google_auth_httplib2.AuthorizedHttp):
+            certs = list(resource._http.http.certificates.iter(""))
+        else:
+            certs = list(resource._http.certificates.iter(""))
+        if has_client_cert:
+            self.assertEqual(len(certs), 1)
+            self.assertEqual(
+                certs[0], (self.ADC_KEY_PATH, self.ADC_CERT_PATH, self.ADC_PASSPHRASE)
+            )
+        else:
+            self.assertEqual(len(certs), 0)
+
+    def client_encrypted_cert_source(self):
+        return self.ADC_CERT_PATH, self.ADC_KEY_PATH, self.ADC_PASSPHRASE
+
+    def test_mtls_not_trigger_if_http_provided(self):
+        discovery = open(datafile("plus.json")).read()
+        plus = build_from_document(discovery, http=httplib2.Http())
+        self.assertTrue(plus is not None)
+        self.assertEqual(plus._baseUrl, self.REGULAR_ENDPOINT)
+        self.check_http_client_cert(plus, has_client_cert=False)
+
+    def test_exception_with_client_cert_source(self):
+        discovery = open(datafile("plus.json")).read()
+        with self.assertRaises(MutualTLSChannelError):
+            build_from_document(
+                discovery,
+                credentials=self.MOCK_CREDENTIALS,
+                client_options={"client_cert_source": mock.Mock()},
+            )
+
+    @parameterized.expand(["Never", "Auto", "Always"])
+    def test_mtls_with_provided_client_cert(self, use_client_cert):
+        discovery = open(datafile("plus.json")).read()
+        os.environ["GOOGLE_API_USE_MTLS"] = use_client_cert
```

Comment:
Please use `patch.dict()` for modifications to the environment variables in tests

https://docs.python.org/3.7/library/unittest.mock.html#patch-dict

Reply:
done